### ISSUE TITLE
FR: Flashcard plugin

### ISSUE THREAD

**Issue by EyalGol** [koreader/koreader#8734]
# Flashcard plugin

## Problem
I think that most people that read nonfiction at some point had the problem of finishing a great book but realizing that it's hard to recall anything specific, even the **highlight**.

## Solution
A Flashcard plugin for Notes and Highlights.
I Invision something along the lines of [Anki](https://apps.ankiweb.net/) integrated into koreader.

It will fetch all notes and highlights from every book in your library and will have the following screens:

![Plugin menu](https://user-images.githubusercontent.com/37214951/151218883-169575f6-764d-44b9-bd30-66bf3780cb45.png)

![Plugin main](https://user-images.githubusercontent.com/37214951/151219530-fa44e697-0bb7-4b00-a156-7934b563ecaf.png)

It will use a spaced repetition algorithm to choose highlights and notes from your library.

### Development
I am willing to develop this plugin, I would be happy if you could link some relevant plugins as examples, I never dove into this project so anything from plugins with similar UI to highlights/notes fetching.

Thoughts?

**Comment by Frenzie** [comment]
Also see https://github.com/koreader/contrib for plugins.

The dictionary/Wikipedia search might have the most similar dialog. Or much simpler, something like http://koreader.rocks/doc/modules/ui.widget.multiconfirmbox.html

(The start dialog is basically just <http://koreader.rocks/doc/modules/ui.widget.confirmbox.html>)

**Comment by EyalGol** [comment]
Thanks :)

I assume you would want me to create a PR into https://github.com/koreader/contrib?

**Comment by hius07** [comment]
Current bookmark details dialog (with any buttons) is displayed here
https://github.com/koreader/koreader/blob/b7a2a27590ba3bf79a5881452b7e548eff74aa4a/frontend/apps/reader/modules/readerbookmark.lua#L537-L597
Highlights and bookmarks are stored separately, but the highlighted text is copied to bookmarks, so there is no need to parse highlights.
Good luck!

**Comment by Frenzie** [comment]
> I assume you would want me to create a PR into https://github.com/koreader/contrib?

In the sense that if you want to share it with others, it seems a bit out of scope for the main repo, yes. Additionally you don't have to worry about our coding style there if you don't want to.

**Comment by EyalGol** [comment]
Okay got it.
I have one question tho, I started working on the plugin everything was going fine and then the plugin stopped being recognized.  it is clearly in the plugins folder but it does not show in the Plugin management menu.
Any idea why?
[flashcard.koplugin.zip](https://github.com/koreader/koreader/files/7946086/flashcard.koplugin.zip)

**Comment by offset-torque** [comment]
Writing this as a long time Anki user. This is a nice idea. But (as you know) you have to implement scheduling algorithm/database and provide at least a basic deck management interface. At that point, this turns into crudely replicating Anki in KOReader.

I don't want to discourage you from contributing but I have to point this out:
**Instead of reinventing Anki, you can write a plugin that can export highlights/notes as an Anki importable CSV.** 

This way:
- You can pass heavy lifting to Anki and save yourself from spending lots of effort replicating the app.
- Flashcard database will be more secure in Anki desktop+mobile sync+Ankiweb backup. I wouldn't want to lose all my note flashcards from all my read books if something happens to my e-reader. In best case, you will instruct users to manually backup deck+scheduling info to their computers.
- Flashcard database won't be locked up. If I sell my e-reader or decide to use a different e-reader software other than KOReader, my flashcard database will be useless. If it is in Anki, I can continue using that on different devices. In case you are planning to make it exportable, remember, scheduling information is as important as the deck and that might be a problem with constant Anki algorithm changes.
- It will (probably) be easier for you to maintain the plugin if it's just an export plugin than a full featured GUI applet inside KOReader.

I mean if you are doing this to pass time / improve your programming skills, or writing this plugin just for yourself, you can ignore things I wrote. But if you are doing this to help users, export approach might have higher effort/benefit ratio. Anki users will probably avoid using this for the reasons I mentioned. So your user base might be much smaller than you expected.

**Comment by Frenzie** [comment]
> At that point, this turns into crudely replicating Anki in KOReader.
> […]
> Instead of reinventing Anki, you can write a plugin that can export highlights/notes as an Anki importable CSV.

That's actually how I interpreted the OP. To write something that can exchange data with Anki including a "crude" (your words ;-) ) implementation of only the review aspect. :-)

**Comment by EyalGol** [comment]
Welp nope, I literally wanted and I [did](https://github.com/EyalGol/contrib) implement flashcards in koreader.
as far as:
> Writing this as a long time Anki user. This is a nice idea. But (as you know) you have to implement a scheduling algorithm/database and provide at least a basic deck management interface. At that point, this turns into crudely replicating Anki in KOReader.

It's not how envisioned it, at least for me making a deck out of all of my highlights is not a good solution, and altho I know that making cards without questions is not the best for learning it's what's practical for me. I just show a card with the quote and 3 buttons (did you remember: no/vaguely/yes) and I use that with a variation of the SM2 algorithm.

The implementation is indeed crude, I didn't try to fully understand the framework and I copied a lot of code (especially the highlight parsing part) but it works and that's what's important for me.

Imma open a PR now, if you test it and find this plugin interesting feel free to review it properly and suggest changes as I'm sure there are plenty.

anyway thanks for the help, this project impresses me over and over with the amount of activity and support it has.

**Comment by offset-torque** [comment]
Oh I see you already implemented it. In that case I change my stance to **"Yeah, why not !"**

I am sure there will be users that don't want to go full Anki but will enjoy reviews. Even just for the purpose of a quick review after finishing the book, this sounds very useful actually (without going into export/transfer/import workflow).

By the way, I used **crude** as a "basic function set implementation". But in your use case, this is enough of course.

> Imma open a PR now, if you test it and find this plugin interesting feel free to review it properly and suggest changes as I'm sure there are plenty.

If this is directed to me, surely I can share my opinions if I think there's room for improvement.
**Best of luck !**

**Comment by hius07** [comment]
You can save a lot of code using standard TextViewer instead of your FlashcardWidget.

**Comment by EyalGol** [comment]
> You can save a lot of code using standard TextViewer instead of your FlashcardWidget.
Thanks I'll look into it.

Also two question:
* Is there a way to spawn a blocking widget (or waiting on it with a "select" after it was spawned)? currently I spaws all of the cards simultaneously and im not proud of it. 
* Is there a way to make a "link" to a book, I would like if cards had a link that wolud open the location of the highlight

**Comment by EyalGol** [comment]
@offset-torque Hi, if you want to check out the plugin it was merged into https://github.com/koreader/contrib

**Comment by offset-torque** [comment]
**Quick notes:**

When I run your plugin (FP), I expected it to bring either:
1. Highlights from the current book
2. Or from the last finished book

But FP brought higlights from one of the books I am currently reading.
Why did it chose that book I don't know.
I repeated the action, it brought again from that book although I have close to 10 books open.

**Current limitations:**
1. I can't choose which book to review.
2. I don't know how many cards are there in a book and how many I reviewed, and what are their status. There is no progress overview.
3. I don't know which books I already reviewed.
4. Close button doesn't end the session, I had to go through all the cards to exit.

Also making user choose how many cards they will review before the session is not intuitive. In very rare cases users can estimate that number properly, in others they have to restart because they want to review more or end the session because they overestimated.

**Suggestion:**
- Start the session without asking a number
- End the session when close button is tapped

**Comment by EyalGol** [comment]
Thanks! I appreciate the feedback.

I agree that I should stop when the exit button is pressed, I dont
understand how to implement it, didnt find a way to have a blocking widget.
And I like your suggested change that I shouldent ask how many cards the
user wants, I think I should ask how much time the user wants to review
cards instead.

Reviewing your progress/book status should be a feature.

Amd about the fact that you got highlights only from a random book. I
currently go sequentially over highlights, the book that was chosen at
seemingly random is your first entry in the history.

**Comment by hius07** [comment]
> didnt find a way to have a blocking widget

Maybe do not show all the cards at once, just one by one?

**Comment by EyalGol** [comment]
> Maybe do not show all the cards at once, just one by one?
Sure but for that I either need a blocking widget or some kind of a "select" on the widget.

I'm 100% sure that there is a way to do it, I'm just not familiar enough with the framework to implement it.

**Comment by hius07** [comment]
Let `displayFlashcards(card)` show one card only, then `displayFlashcardCallback` will call `displayFlashcards(nextcard)`.

**Comment by EyalGol** [comment]
> Let displayFlashcards(card) show one card only, then displayFlashcardCallback will call displayFlashcards(nextcard).

I could but it would need to be recursion where showFlashcard would get (highlights, index) as its arguments. I really dont like this implementation :/

Is that the only way to implement such a thing? (There are no feature to implement a blocking UI element in any way?) @Frenzie @hius07

**Comment by Frenzie** [comment]
I'm not quite sure what you mean by blocking in this context. You can set modal=true but as @hius07 said it sounds like if you're rendering 50 widgets all at once that's the problem and you just shouldn't be rendering 50 widgets all at once.

> I could but it would need to be recursion where showFlashcard would get (highlights, index) as its arguments. I really dont like this implementation :/

You don't need to pass them around as arguments — well, possibly the index, but not the highlights anyway. The idiomatic way would normally be self.highlights.

**Comment by poire-z** [comment]
There would be no need for recursion.
You shouldn't need anything blocking, that's not how KOReader does things: it is event driven, so you show a widget, and you're done. You set callbacks (ok_callback, cancel_callback, dismiss_callback...) to this widget so it can go on doing things when an event happen.
Haven't looked at your code, but from what you say, you could put all your items in a table array `table.insert(self.all_cards some_item)`, have a showCard() that pop a card from that array `card_to_show=table.remove(self.all_cards some)`, and in the next_callback of that widget, just `UIManager.close(that_widget) ; showCard()` to show the new one that will pick a new card_to_show from your all_cards.

(We have a thing for blocking linear Q/A dialogs sequences, frontent/ui/trapper.lua, used for Wikipedia save as EPUB or Extract cached book information, but it is a bit complex and feels overkill for what you want to do.)

**Comment by EyalGol** [comment]
> There would be no need for recursion.
You shouldn't need anything blocking, that's not how KOReader does things: it is event driven, so you show a widget, and you're done. You set callbacks (ok_callback, cancel_callback, dismiss_callback...) to this widget so it can go on doing things when an event happen.
Haven't looked at your code, but from what you say, you could put all your items in a table array table.insert(self.all_cards some_item), have a showCard() that pop a card from that array card_to_show=table.remove(self.all_cards some), and in the next_callback of that widget, just UIManager.close(that_widget) ; showCard() to show the new one that will pick a new card_to_show from your all_cards.

Sounds good, I'm just so used to C and method callbacks are weird for me. Thanks for the help.

My table is a bit more complex so I'll need a custom pop function but thats about it.

**Comment by costis-t** [comment]
... language learning maniac here. May I propose a future Feature Request for this great idea:

- Put each single-word highlight of a book on one page of the flashcard
- its Translate definition or Dictionary or both on the other.

**Comment by EyalGol** [comment]
> ... language learning maniac here. May I propose a future Feature Request for this great idea:

> Put each single-word highlight of a book on one page of the flashcard

> its Translate definition or Dictionary or both on the other.

Cool! Didnt even think about this application, not exactly my focus atm but definitely could be a toggle a toggled in the settings! 
Like another alternative mode that will turn the card into language learning cards or even a better idea, in koreader there are different types (colors) of highlights, I could use that to determine if its a language highlights.

@costis-t How do you envision the rating for language cards looks like?
* I could ask you to write the translation, but for me writing a translation on an eink keyboard sound like pain. 
* I could have multiple answers but I dont know where I could get them from (if not from the user himself, which I could do, lets say that for a card to be displayed as a language card you would need to add a note with the folowing format "t: thing, thing, thing") 
* or to leave it with the existing 3 buttons :)

**Comment by costis-t** [comment]
I plan on working on something similar in March so I do not have any concrete ideas right now, but wow, you also did the "SuperMemo" algorithm, OK! :)

Yes, having a rich settings menu is cool for various use cases.

Having reviewed the above, I believe a feature-rich export to Anki would be just fine. But if you want to experiment with an implementation in the fantastic minimalism of koreader, I describe my current vision.

1. Take the one-word (or max two-word?) highlights from an ebook.
2. Create the databases with the word definitions using google-translate (the Translator, `frontend/ui/translator.lua`)
     - potentially also store the sentence/phrase/nearby words (I don't know which one is optimal in practice) that contains the highlighted word. This will help the learner take some context in case (s)he nearly remembers the meaning of the word (and also pave stronger neural pathways, I presume). 
3. Create the cards. Two-sided is fine for my current interest in western languages, but for others languages like Hebrew/Chinese adding a pronunciation-side might also be helpful like [wikipedia](https://en.wikipedia.org/wiki/Flashcard) describes. Small steps, though, and we build on them.
4. Give the option to reverse the Question/Answer for the good learners. Question: the google-translated text, Answer: the original highlight.

![Screenshot_20220131_104332](https://user-images.githubusercontent.com/20000896/151772444-02e28c10-7bc8-4335-aee3-7210a9952225.png)

**Comment by EyalGol** [comment]
> I plan on working on something similar in March so I do not have any concrete ideas right now, but wow, you also did the "SuperMemo" algorithm, OK! :)
> 
> Yes, having a rich settings menu is cool for various use cases.
> 
> Having reviewed the above, I believe a feature-rich export to Anki would be just fine. But if you want to experiment with an implementation in the fantastic minimalism of koreader, I describe my current vision.
> 
> 1. Take the one-word (or max two-word?) highlights from an ebook.
> 2. Create the databases with the word definitions using google-translate (the Translator, `frontend/ui/translator.lua`)
>      - potentially also store the sentence/phrase/nearby words (I don't know which one is optimal in practice) that contains the highlighted word. This will help the learner take some context in case (s)he nearly remembers the meaning of the word (and also pave stronger neural pathways, I presume). 
> 3. Create the cards. Two-sided is fine for my current interest in western languages, but for others languages like Hebrew/Chinese adding a pronunciation-side might also be helpful like [wikipedia](https://en.wikipedia.org/wiki/Flashcard) describes. Small steps, though, and we build on them.
> 4. Give the option to reverse the Question/Answer for the good learners. Question: the google-translated text, Answer: the original highlight.
> 
> ![Screenshot_20220131_104332](https://user-images.githubusercontent.com/20000896/151772444-02e28c10-7bc8-4335-aee3-7210a9952225.png)
> 

Thanks for the feedback :)
As for the current "vision", I think its pretty easy to implement.
For starters I can just use the existing translation view and just show it after every card if the card text is less then X words.

fyi, made this plugin in 2 days without prior knowledge in lue or the koreader codebase. Looks like there are people interested in this plugin so I think I'll remake it properly so it'll support expention. @costis-t if you want to implement the "language" cards I think it could be great, anyway I think I'll have a basic shot on it.

**Comment by costis-t** [comment]
@EyalGol koreader seems a very appetizing project to built on, however I have these hopefully only temporary obstacles: my February is really full and I hope I get some time in March, I'd like to cover a lot of my OOP gaps before working on it, I also have no clue about lua/koreader-codebase but what you said is hopeful!

**Comment by goobta** [comment]
@EyalGol thank you for your work! I recently discovered KOReader / Kobo, and I love the idea of being able to develop on an e-screen. I'm thinking about applying to grad school, and thought a great use for this would be to learn my GRE words at night before bed. I appreciate that you already implemented spaced repetition, so most of the heavy lifting is (hopefully) already done.

I'm wondering what's the best way to sideload / bulk load words into the database. There are two big questions that would probably need consensus on
1. The UI (do users load a file? What's the best format?)
2. If it's possible to store "definitions" alongside words. I think that should be easy enough; but then the wammy followup: can we display rich text definition? i.e. I have formatting / photos for a bunch of my words (from Anki); I'd love if we could display HTML or something similar on the screen.

I guess that you can consider this a FR. I'm not familiar with Lua / KOReader, but you mentioned earlier that it's easy to get started. You also stated that you got ramped up pretty fast, so I might be able to do this myself with a few pointers. However, if you're redoing the plugin, I don't want to step on your toes-- I'd much rather we combine forces than someone getting very frustrated.
